src/OptimizedPathMappingRepository.php

```diff
@@ -0,0 +1,365 @@
+
+        // Run the assertion after find(), so that we know that $query is valid
+        Assert::notEq('', trim($query, '/'), 'The root directory cannot be removed.');
+
+        foreach ($resources as $resource) {
+            $this->removeResource($resource);
+        }
+
+        return $nbOfResources - count($this->store->keys());
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function clear()
+    {
+        $root = new GenericFilesystemResource('/');
+        $root->attachTo($this, '/');
+
+        // Subtract root
+        $removed = count($this->store->keys()) - 1;
+
+        $this->store->clear();
+        $this->store->set('/', $root);
+
+        return $removed;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function listChildren($path)
+    {
+        $iterator = $this->getChildIterator($this->get($path));
+        $children = $this->iteratorToCollection($iterator);
+
+        return new FilesystemResourceCollection($children);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function hasChildren($path)
+    {
+        $iterator = $this->getChildIterator($this->get($path));
+        $iterator->rewind();
+
+        return $iterator->valid();
+    }
+
+
+    /**
+     * @param string $path
+     * @param FilesystemResource $resource
+     */
+    private function addResource($path, FilesystemResource $resource)
+    {
+        // Don't modify resources attached to other repositories
+        if ($resource->isAttached()) {
+            $resource = clone $resource;
+        }
+
+        // Read children before attaching the resource to this repository
+        $children = $resource->listChildren();
+
+        $resource->attachTo($this, $path);
+
+        // Add the resource before adding its children, so that the array
+        // stays sorted
+        $this->store->set($path, $resource);
+
+        $basePath = '/' === $path ? $path : $path.'/';
+
+        foreach ($children as $name => $child) {
+            $this->addResource($basePath.$name, $child);
+        }
+    }
+
+
+    /**
+     * @param FilesystemResource $resource
+     */
+    private function removeResource(FilesystemResource $resource)
+    {
+        $path = $resource->getPath();
+
+        // Ignore non-existing resources
+        if (! $this->store->exists($path)) {
+            return;
+        }
+
+        // Recursively register directory contents
+        foreach ($this->iteratorToCollection($this->getChildIterator($resource)) as $child) {
+            $this->removeResource($child);
+        }
+
+        $this->store->remove($path);
+
+        // Detach from locator
+        $resource->detach($this);
+    }
+
+    /**
+     * Returns an iterator for the children of a resource.
+     *
+     * @param FilesystemResource $resource The resource.
+     *
+     * @return RegexFilterIterator|Resource[] The iterator.
+     */
+    private function getChildIterator(FilesystemResource $resource)
```

Comment:
This function (and `getGlobterator()`) returns keys where equivalent in other repositories returns resources. It does not seem to be a problem as the method is private, but is it?